src/main/java/com/faforever/client/io/DownloadService.java

```diff
@@ -1,49 +1,119 @@
 package com.faforever.client.io;
 
+import com.faforever.client.preferences.PreferencesService;
 import com.faforever.client.task.ResourceLocks;
 import com.faforever.commons.io.ByteCopier;
 import com.faforever.commons.io.ByteCountListener;
+import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 import org.springframework.context.annotation.Lazy;
 import org.springframework.stereotype.Service;
 
+import javax.xml.bind.DatatypeConverter;
+import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URI;
 import java.net.URL;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
 @Service
 @Lazy
 @Slf4j
+@RequiredArgsConstructor
 public class DownloadService {
+  private final PreferencesService preferencesService;
 
-  public void downloadFile(URL url, Path targetFile, ByteCountListener progressListener) throws IOException {
+  /*
+   * Download file first trying each mirror in order. If all mirrors fail then try the original URL.
+   */
+  public void downloadFileWithMirrors(URL url, Path targetFile, ByteCountListener progressListener, String md5sum) throws IOException, NoSuchAlgorithmException, ChecksumMismatchException {
+    for (URL mirrorUrl : getMirrorsFor(url)) {
+      try {
+        downloadFile(mirrorUrl, targetFile, progressListener, md5sum);
+        return;
+      } catch (FileNotFoundException e) {
+        log.warn("Could not download file at {}", mirrorUrl);
```

Comment:
log the exception here please. Also this can be combined with the IOexception catch no need to have 2 that are the same

Reply:
Oh yea the weird thing here is that URL will throw this `FileNotFoundException` if it encounters an http error when trying to open the URL. That's why I separated it out, because it's specifically catching the http errors. I should probably make it log on INFO or something though because I would expect this to happen all the time given that the mirrors may be out of date and not contain all needed files.

Reply:
Maybe but if the error message and log levels are the same then there is no need to separate them out since they are performing the same action

Reply:
I changed it to info